Revise section "Exceptions". (#596)
zopefoundation/Zope: docs/zdgbook/ObjectPublishing.rst

Exceptions
----------

When the object publisher catches an unhandled exception, it tries to
match it with a set of predifined exceptions coming from the
Products.zExceptions package, such as **HTTPNoContent**, **HTTPNotFound**,
**HTTPUnauthorized** and many, many more. If there is a match, the custom

> exception gets upgraded into an zException, which then results in a proper

Reviewer: upgraded **to a** zException

Instead of punting by just saying "many, many more" I would list them like the old documentation did.

Author: I removed the "many, many more" phrase and added a link to the package for a full list of exceptions..

The old documentation included more, but also only a partial list of all the possible exceptions.

Plus, if there will be a change to the list of exceptions of the zExceptions package (which probably won't ever happen), nobody will think about updating the Zope developer guide.

Also, if I really want to implement a custom exception I am probably more interested to choose the right status code. Choosing the name is only a way to get there. Either way, I'd have to lookup the source code of zExceptions anyway.

response returned to the browser, including a HTTP status code.

.. attention::

    When you create a custom exception, please make sure not to inherit
    from **BaseException**, but e.g. from **Exception**, otherwise you'll
    run into an waitress exception.

.. note::

    Beginning with Zope 4, a standard installation no longer comes with
    a ``standard_error_message``.

    There are two ways to catch and render an exception:

    - create a ``standard_error_message``, which can be a **DTML Method**, **DTML Document**, **Python Script** or **Page Template**
    - create an ``exception view``, see blog post `Catching and rendering exceptions <https://blog.gocept.com/2017/10/24/zope4-errorhandling/>`_

If the exception is not handled, it travels up the WSGI stack.

What then happens depends entirely on the WSGI middleware. By default
Zope comes with **waitress** and by default **waitress** returns an error
message as follows::

  Internal Server Error

  The server encountered an unexpected internal server error

  (generated by waitress)

> .. note::

Reviewer: Instead of listing particulars here just point to this: https://zope.readthedocs.io/en/latest/wsgi.html#debugging-zope-applications-under-wsgi

Author: Right! Added a link to the **debugging Zope applications under WSGI** documentation. Btw - thanks a lot for writing it! There was really a need for it.

Exceptions
----------

When the object publisher catches an unhandled exception, it tries to
match it with a set of predifined exceptions coming from the
**zExceptions** package, such as **HTTPNoContent**, **HTTPNotFound**,
**HTTPUnauthorized**.

> If there is a match, the exception gets upgraded into the matching

Reviewer: upgraded **to**

Author: Thanks for taking care even of those details!

**zException**, which then results in a proper response returned to the

> browser, including a HTTP status code.

Reviewer 2: browser, including an appropriate HTTP status code.

.. note::

     For a full list of exceptions please directly refer to the
     implemented exception classes within the
     `zExceptions package
     <https://github.com/zopefoundation/zExceptions/blob/master/src/zExceptions/__init__.py>`_.

.. attention::

    When you create a custom exception, please make sure not to inherit

>     from **BaseException**, but e.g. from **Exception**, otherwise you'll

Reviewer 3: ```suggestion
    from **BaseException**, but from **Exception** or one of its child classes, otherwise you'll
```

>     run into an waitress exception.

Reviewer 3: ```suggestion
    run into an exception in waitress.
```

.. note::

    Beginning with Zope 4, a standard installation no longer comes with
    a ``standard_error_message``.

    There are two ways to catch and render an exception:

>     - create a ``standard_error_message``, which can be a **DTML Method**, **DTML Document**, **Python Script** or **Page Template**

Reviewer 3: ```suggestion
    - create a ``standard_error_message``, which can be a **DTML Method**, **DTML Document**, **Script (Python)** or **Page Template**
```

Changed to the name used in the add types menu in the ZMI.

    - create an ``exception view``, see blog post `Catching and rendering exceptions <https://blog.gocept.com/2017/10/24/zope4-errorhandling/>`_

If the exception is not handled, it travels up the WSGI stack.

> What then happens depends entirely on the WSGI middleware. By default

Reviewer: **happens then** instead of then happens

Author: Thanks for taking care even of those details!

> What happens then depends entirely on the WSGI middleware. By default

Reviewer 3: ```suggestion
What happens then depends entirely on the possibly installed WSGI middleware or the WSGI server. By default
```

> Zope comes with **waitress** and by default **waitress** returns an error

Reviewer 3: ```suggestion
Zope by default uses **waitress** which returns an error
```
